File: ghc-4335c07/nofib/real/fem/inC/displacement.c
/* compute structural displacement and print them out */

#include <stdio.h>
#include "database"
extern FILE *fp;


extern int diagadr[];
extern float  stiff_vbm[], load_vec[];

getnuvw(uvw,node,nuvw)

int node;
float  uvw[],nuvw[];          

/* build up the displacement array of a node */
{ 
  int dgr;
  float  u,v,theta;

  dgr = node_info[node].u;
  if (dgr!=0) 
     u = uvw[dgr];
  else 
     u = 0.0;

  dgr = node_info[node].v;
  if (dgr!=0)
     v = uvw[dgr];
  else
     v = 0.0;

  dgr = node_info[node].theta;
  if (dgr!=0)
     theta = uvw[dgr];
  else
     theta = 0.0;

  nuvw[0] = u;
  nuvw[1] = v;
  nuvw[2] = theta;

}

displacement() {

  int node, bc;
  float  x,y,u,v,theta,nuvw[3];

  vblldecomp(ndgrs,diagadr,stiff_vbm);
  vbllsolution(ndgrs,diagadr,stiff_vbm,load_vec);

  printf("\n\nDISPLACEMENT OF THE STRUCTURE \n\n");
  printf(" Node      X             Y       BC        U             V");
  printf("           Theta\n");

  for (node=1; node<=nnode; node=node+1) {

    x = node_info[node].x;
    y = node_info[node].y;
    bc= node_info[node].bc;
    getnuvw(load_vec,node,nuvw);
    u = nuvw[0];
    v = nuvw[1];
    theta = nuvw[2];

    printf("%4d%14.6f%14.6f%4d%14.6f%14.6f%14.6f\n",node,x,y,bc,u,v,theta);
    fprintf(fp,"%14.6f   %14.6f\n",u,v);

  }

}

File: ghc-4335c07/nofib/real/fem/Displacement.hs
-- Glasow Haskell 0.403 : FINITE ELEMENT PROGRAM V2
-- **********************************************************************
-- *                                                                    *
-- * FILE NAME : displacement.hs              DATE : 13-3-1991          *
-- *                                                                    *
-- * CONTENTS : Compute nodal displacement of the structure.            *
-- *                                                                    *
-- **********************************************************************

module Displacement ( uvw, getnuvw ) where

import Basics
import Vector
import Matrix
import VBmatrix
import VBlldecomp
import DB_interface
import Degrees
import Pre_assemble
import Assemble_stiffness
import Assemble_loadvec

uvw     :: (Array Int Int, Array Int Float) -> Vec Float

getnuvw :: (Array Int Int, Array Int Float) -> Int -> 
	    	Vec Float -> (Float, Float, Float)

t_Ub s = vbllsolution (kdd s) (loadvec s)

uvw s =
	incrvec initial_value index_value_assoc
	where
	initial_value = makevec ( 3 * (nnode s) ) ( \ i -> 0.0 )
	index_value_assoc = concat (map f_s [1..(nnode s)])
	f_s = f s tUb
	tUb = t_Ub s

f s tUb node =
	azip [l,l+1,l+2] (map ff dgrs)
	where
	l = 3 * (node - 1) + 1
	dgrs = getndgr s node
	ff i = if ( i == 0 ) then 0.0 else vecsub tUb i

getnuvw s node uvw =
	(u,v,theta)
	where
	u = vecsub uvw index
	v = vecsub uvw (index+1)
	theta = vecsub uvw (index+2)
	index = 3 * (node - 1) + 1 

